dev/spark_overview.md

```diff
@@ -0,0 +1,452 @@
+- When there is really only enough work for one executor but Spark distributes it to all executors
+anyway. The inter-process communication adds tons of overhead. The executors also appear to do a
+lot more work. With dynamic allocation Spark gradually adds executors and these problems don't
+always occur.
+
+The second issue sometimes occurs when submitting a dataset to a project in the registry.
+We recommend enabling dynamic allocation when doing this. If that is enabled then dsgrid code will
+reconfigure the SparkSession to use a single executor with one core.
+
+Set these values in your `spark-defaults.conf`. `spark.shuffle.service.enabled` must be set before
+you start the workers.
+```
+spark.dynamicAllocation.enabled true
+spark.dynamicAllocation.shuffleTracking.enabled true
+spark.shuffle.service.enabled true
+```
+
+### Slow local storage
+Spark will write lots of temporary data to local storage during shuffle operations. If your joins
+cause lots of shuffling, it is very important that your local storage be fast. If you use an Eagle
+node with a slow spinning disk (130 MB/s), your job will likelly fail. Use a `bigmem` or `gpu`
+node. They have SSDs that can write data at 2 GB/s.
+
+We have not observed any downside to having this feature enabled.
+
+### Too many executors are involved in a job
+Some of our Spark jobs, particularly those that create large, in-memory tables from dimension
+records and mappings, perform much better when there is only one executor and only one CPU. If
+you think this is happening then set these values in your `spark-default.confg`:
+```
+spark.executor.cores 1
+# This value should be greater than half of the memory allocated to the Spark workers, which
+# will ensure that Spark can only create one executor.
+spark.executor.memory 15g
+```
+One common symptom of this type of problem is that a job run in local mode works better than in a
+standalone cluster. The likely reason is that the standalone cluster has many executors and local
+mode only has one.
+
+### A table has partitions that are too small or too large.
+Spark documentation recommends that Parquet partition files be in the range of 100-200 MiB, with
+128 MiB being ideal. A very high compression ratio may change that. This affects how much data
+each task reads from storage into memory.
+
+Check your partition sizes with this command:
+```
+$ find <path_to_top_level_data.parquet> -name "*.parquet" -exec ls -lh {} +
+```
+
+If you are not partitioning by column and know the target number of partitions:
+
+If there are too few partitions currently:
+```
+df.repartition(X).write.parquet("data.parquet")
+```
+
+If there are too many partitions currently:
+```
+df.coalesce(X).write.parquet("data.parquet")
+```
+
+If you are partitioning by a column and find that there are many very small partition files,
+repartition like this:
+```
+df.repartition(column_name).write.partitionBy(column_name).parquet("data.parquet")
+```
```

Comment:
Is it correct that there is no guidance in this call as to how many partitions to create or how large the partitions should be? Is `repartition` doing the work of grouping the columns differently?

Reply:
Above I said the guidance was 100-200 MiB. I'll add a `du -sh` command to get the total size.

Reply:
The text above recommends a target size range.